File: House_robber_recursion.cpp
//recursion
class Solution {
public:
    int rob(vector<int>& nums) {
     return helper(nums, nums.size() - 1);
    }

    int helper(vector<int>& nums, int i){
        //base case
        if(i >= nums.size() ) return 0;

        // case
        return std::max(helper(nums, i-2) + nums[i], helper(nums, i-1));
    }


};

//Time complexity: O(2^n)
//Leet Code: Time limit exceeded

File: HouseRobber_Recursion.java
/**
Leet Code Submitted : NA (Since,its a recursion based solution and will cause Time Limit Exceeded Error)
Space Complexity : O(2^N)
Time Complexity : O(2^N)
The idea is to recursive try all the sub-problems in decision tree. The decision will be to rob the particular house or not**/

class Solution {
    public int rob(int[] nums) {
        
        int max = 0;
        if(nums == null || nums.length < 1)
            return max;
        return decision(nums,0,max) ; 
    }
    
    int decision(int[] nums,int i,int max){
        if(i >= nums.length)
            return max;
    
        return (Math.max(decision(nums,i+2,max+nums[i]),decision(nums,i+1,max)));
        
    }
}